scripts/xtensa-build-zephyr.py

```diff
@@ -0,0 +1,370 @@
+						"created and west will recognize it as its new sof module."
+						"If zephyr-project/modules/audio/sof already exists and is a"
+						"different copy than where this script is run from, then the"
+						"behavior is undefined."
+						"This -p option is always required_if the real (not symbolic)"
+						"sof/ and zephyr-project/ directories are not nested in one"
+						"another.")
+	mode_group.add_argument("-c", "--clone", required=False, action="store_true",
+						help="Using west, downloads inside this sof clone a new Zephyr"
+						"project with the required git repos. Creates a"
+						"sof/zephyrproject/modules/audio/sof symbolic link pointing"
+						"back at this sof clone."
+						"Incompatible with -p. To stop after downloading Zephyr, do not"
+						"pass any platform or cmake argument.")
+	parser.add_argument("-v", "--verbose-build", required=False, action="store_true",
+						help="Verbose build system log")
+	parser.add_argument("-C", "--cmake-args", required=False,
+						help="Cmake arguments passed to cmake configure step without\n"
+						"'-D' prefix. Eg. CMAKE_GENERATOR='Ninja'. Preserves internal quotes.")
+
+	args = parser.parse_args()
+
+	if args.all:
+		args.platforms = platform_names
+
+	# print help message if no arguments provided
+	if len(sys.argv)==1:
+			parser.print_help()
+			sys.exit(0)
+
+	if args.z and not args.clone:
+		raise RuntimeError(f"Argument -z without -c makes no sense")
+	if args.clone and not args.z:
+		args.z = "main"		# set default name for -z if -c specified
+
+	if args.west_path:
+		WEST_TOP=pathlib.Path(args.west_path)
+
+	if args.verbose_build:
+		if not args.cmake_args:
+			args.cmake_args = "CMAKE_VERBOSE_MAKEFILE=ON"
+		else:
+			args.cmake_args = "CMAKE_VERBOSE_MAKEFILE=ON " + args.cmake_args
+
+	# split arguments by whitespaces and append -D prefix to match expected CMake format
+	if args.cmake_args:
+		arg_list = shlex.split(args.cmake_args)
+		arg_list = ["-D" + arg for arg in arg_list]
+		args.cmake_args = " ".join(arg_list)
+
+def git_submodules_update():
+	global SOF_TOP
+	subprocess.run("git submodule update --init --recursive", shell=True, timeout=600,
+		check=True, cwd=SOF_TOP)
+
+def check_west_installation():
+	west_path = shutil.which("west")
+	if not west_path:
+		raise FileNotFoundError("Install west and a west toolchain,"
+			"https://docs.zephyrproject.org/latest/getting_started/index.html")
+	print(f"Found west: {west_path}")
+
+def find_west_workspace():
+	"""[summary] Scans for west workspaces using west topdir command by checking script execution
+	WEST_TOP and SOF_TOP directories.
+
+	:return: [description] Returns path when west workspace had been found, otherwise None
+	:rtype: [type] bool, pathlib.Path
+	"""
+	global WEST_TOP, SOF_TOP
+	paths = [ pathlib.Path(os.getcwd()), SOF_TOP, WEST_TOP]
+	for path in paths:
+		if path.is_dir():
+			result = subprocess.run("west topdir", shell=True, timeout=5, capture_output=True,
+				check=False, cwd=path)
```

Comment:
Can you please provide some reference on the capture_output deadlock problem? I could not find any.

Reply:
This will not happen as long as `west topdir` output volume is small - smaller than the buffer size of the PIPE created by `subprocess`. However once that (unknown) line is crossed then deadlock is guaranteed: `west topdir` is blocked waiting for the pipe to have more space. Meanwhile, this script is blocked waiting for the `west topdir` exit code.

https://docs.python.org/3/library/subprocess.html#popen-objects
https://www.google.com/search?q=subprocess+communicate+deadlock

Reply:
> To discard the output use stdout=DEVNULL and stderr=DEVNULL

Actually: use `stdout=DEVNULL` only let and stderr pass through (the default). Never discard stderr...

Reply:
Well, I think that you may be wrong here.
First, in the[ subprocess.run](https://docs.python.org/3/library/subprocess.html#subprocess.run) documentation they stated the following: 
> Note The newlines attribute of the file objects Popen.stdin, Popen.stdout and Popen.stderr are not updated by the Popen.communicate() method.

The link you send me https://docs.python.org/3/library/subprocess.html#popen-objects has that deadlock warning under Popen.wait() function description.
The first googling result I find is [this thread](https://stackoverflow.com/questions/14615462/why-does-communicate-deadlock-when-used-with-multiple-popen-subprocesses) seems to be related to the issue that occurred in older python versions. And this [fix](http://bugs.python.org/issue12786) seems to be the resolution but I am unsure.
But whatever it does not matter really, I will devnull them.